### ISSUE TITLE
Multiple potential security issues

### ISSUE THREAD

**Issue by klausi** [slootjes/drupal-example-api-module#1]
Hi, I saw your Drupalcon video about this module, very interesting and I learned a lot!

I tried to get this module running, but I always get a 403 JSON error at http://drupal-8.ddev.local/api "error.http.cacheable_access_denied" (Drupal 8.7 dev). Anything I need to do to get this working?

I have a couple of potential security issues I see in the code:

1. Example module should always be secure by default, so there should be no dangerous "@Security(access=true)" that people copy.
2. For listing articles the "access content" permission should be used, which means "View published content". Same for the show action of an individual article.
3. ArticleTransformer does not respect field access, it always just outputs all the fields.
4. The create action should use the permission  for "Article: Create new content".
5. handleForm() looks vulnerable to CSRF attacks. Could not test it because the module does not work for me. Do Symfony forms add CSRF protection by default?
6. Same field access problems when saving the node, we need to check if the user is allowed to write a specific field.

This is probably a Drupal vs. Symfony culture issue: in Symfony you write all your code on your own. The code does not integrate with anything else, and you make assumptions from scratch about it. In Drupal the entity system is deeply connected to everything else and access control is important to not open attack vectors.

For example: users assume that unpublished node articles are not visible anywhere. With your example module they leak through your controller.

**Comment by slootjes** [comment]
Hi, thanks for taking the time to put this together :)

1/2. This module just shows what options are available; this is intentional
3/4. It's not the purpose of this example module to deal with that
5. Not by default; if you need CSRF protection you can implement it using: https://symfony.com/doc/current/security/csrf.html
6. It's not the purpose of this example module to deal with that, it just showcases you can use Symfony Form component in Drupal just fine :)

Please consider this module as a **proof of concept** only; _not_ something you should put in production right away without doing your own development. It is only created with the purpose to show some of this things you can do with controller annotations and some Symfony components within Drupal.

TLDR; yes, your concerns are all valid however I will not address them for the time being as this is not the intention of this example module.

**Comment by klausi** [comment]
The problem is that people copy example code all the time and do not modify it. So it is critically important that example code is secure by default because otherwise people open up vulnerabilities on their projects.

Maybe add to the README: "Warning: the examples shown here are all insecure by default. Please modify them to check properly for permissions/access control/CSRF before putting them into production."

**Comment by slootjes** [comment]
I've made the disclaimer even more explicit than it previously was. Thanks again for taking the time to share your concern 👍
